### PR TITLE
[Smotrim] Add extractor

### LINKED ISSUE #4110: [RUTV] https://smotrim.ru is broken now
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.05.18** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Russian Federation

### Description

The site https://smotrim.ru has become inoperative after changing the advertising policy, which occurred on 06/18/2022.

### Verbose log

```shell
[debug] Command-line config: ['-vU', '-f', '1200', '-o', 'other\\%(upload_date)s_%(title)s.%(ext)s', '-a', 'listurls.txt']
[debug] Batch file urls: ['https://smotrim.ru/video/2427602']
[debug] Encodings: locale cp1251, fs utf-8, pref cp1251, out utf-8 (No ANSI), error utf-8 (No ANSI), screen utf-8 (No ANSI)
[debug] yt-dlp version 2022.05.18 [b14d523] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-7-6.1.7601-SP1
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: ffmpeg -bsfs
[debug] exe versions: ffmpeg 2022-02-07-git-04cc7a5548-full_build-www.gyan.dev (setts), ffprobe 2022-02-07-git-04cc7a5548-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
Latest version: 2022.05.18, Current version: 2022.05.18
yt-dlp is up to date (2022.05.18)
[debug] [generic] Extracting URL: https://smotrim.ru/video/2427602
[generic] 2427602: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] 2427602: Downloading webpage
[generic] 2427602: Extracting information
[debug] Looking for video embeds
ERROR: Unsupported URL: https://smotrim.ru/video/2427602
Traceback (most recent call last):
  File "yt_dlp\YoutubeDL.py", line 1395, in wrapper
  File "yt_dlp\YoutubeDL.py", line 1465, in __extract_info
  File "yt_dlp\extractor\common.py", line 642, in extract
  File "yt_dlp\extractor\generic.py", line 4031, in _real_extract
yt_dlp.utils.UnsupportedError: Unsupported URL: https://smotrim.ru/video/2427602
```

### PR DESCRIPTION
**Pull request by nikita-moor**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Add new extractor for [smotrim.ru](https://smotrim.ru/).

</details>

This video hosting is not supported by the current version of yt-dlp. Perhaps, [\[rutv\]](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/extractor/rutv.py) was somehow related to smotrim.ru (old version of the website?), but now it cannot download video.

Fixes #4110

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by nikita-moor** [comment]
> Read this again

Is it better now?

File: yt_dlp/extractor/smotrim.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            }
+        }, {  # w/o video
+            'only_matching': True,
+            'url': 'https://smotrim.ru/article/2909569',
+        }
+    ]
+
+    def _real_extract(self, url):
+        webpage = self._download_webpage(url, None, 'Downloading webpage')
+        if re.match(self._VALID_URL, url).group('type') == 'brand':
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:61]: self._match_valid_url

File: yt_dlp/extractor/smotrim.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            video_id = self._search_regex(
+                r'"https://player.smotrim.ru/iframe/video/id/(?P<video_id>[0-9]+)/',
+                webpage, 'video_id', default=None)
+            if video_id is None:
+                raise ExtractorError('This page doesn\'t contain video.', expected=True)
+            webpage = self._download_webpage('https://smotrim.ru/video/' + video_id, None, 'Redirect to video')
+
+        # Example: https://player.smotrim.ru/iframe/video/id/1539617/start_zoom/true/showZoomBtn/false/sid/smotrim/isPlay/true/mute/true/?acc_video_id=1488382"
+        m = re.search(
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:70]: self._search_regex

File: yt_dlp/extractor/smotrim.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            r'"https://player.smotrim.ru/iframe/video/id/(?P<video_id>[0-9]+)/[^"]*?/\?acc_video_id=(?P<player_id>[0-9]+)"',
+            webpage)
+        if m is None:
+            raise ExtractorError('This page doesn\'t contain video.', expected=True)
+        video_id = m.group('video_id')
+        player_id = m.group('player_id')
+        player_id_split = '/'.join(re.findall('...', player_id.zfill(9)))  # 2624356 > 002/624/356
+
+        meta = self._download_json(f'https://player.smotrim.ru/iframe/datavideo/id/{video_id}/sid/smotrim', None,)
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:79]: video_id is not None

File: yt_dlp/extractor/smotrim.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        meta = traverse_obj(meta, ('data', 'playlist', 'medialist'))[0]
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:80]: Is this meant to be fatal or not?

File: yt_dlp/extractor/smotrim.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            '360': 'medium-wide',
+            '528': 'high',
+            '540': 'high-wide',
+            '720': 'hd-wide',
+            '1080': 'fhd-wide',
+        }
+        video_template = 'https://cdn-v.rtr-vesti.ru/_cdn_auth/secure/v/vh/mp4/{quality}/{id_split}.mp4?auth=mh&vid={id}'
+
+        formats = []
+        if meta.get('errors') == '':
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:94]:
```suggestion
        if not meta.get('errors'):
```

File: yt_dlp/extractor/smotrim.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            for key in traverse_obj(meta, ('sources', 'http')):
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:95]: again, despite traverse_obj, the code is actually fatal. Decide which one it should be

File: yt_dlp/extractor/smotrim.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+                        'acodec': 'AAC',
+                        'vcodec': 'H264',
+                        'url': video_template.format(quality=video_quality[key], id_split=player_id_split, id=player_id)
+                    })
+                else:
+                    print(f'\nWARNING: unknown resolution "{key}p".')
+                    print('Please, inform the developers. Open a ticket here:')
+                    print('https://github.com/yt-dlp/yt-dlp/issues')
+                    print('Title: [Smotrim] enhance')
+                    print(f'Text: New resolution "{key}" in the video {url}.\n')
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:110]: Use self._report_warning and bug_reports_message

File: yt_dlp/extractor/smotrim.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+                    formats.append({
+                        'format': 'unknown',
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:112]: Use `key`

File: yt_dlp/extractor/smotrim.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+                        'resolution': key,
+                        'ext': 'mp4',
+                        'acodec': 'AAC',
+                        'vcodec': 'H264',
+                        'url': traverse_obj(meta, ('sources', 'http', key))
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:117]: Why not use this when key is known too?

File: yt_dlp/extractor/smotrim.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+                    })
+        elif meta.get('errors') == 'Просмотр видео ограничен в вашем регионе':  # geo-restricted
+            print('Video isn\'t available in your region. Trying to identify accessible formats...')
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:120]: self.to_screen or self._report_warning

File: yt_dlp/extractor/smotrim.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+            for key in video_quality:
+                video_url = video_template.format(quality=video_quality[key], id_split=player_id_split, id=player_id)
+                try:
+                    print(f'   checking {key}p... ', end='')
+                    urllib.request.urlopen(video_url)
+                except urllib.error.HTTPError:
+                    print('fail')
+                    continue
+                print("success")
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:129]: self._check_formats

File: yt_dlp/extractor/smotrim.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+                formats.append({
+                    'format': video_quality[key],
+                    'resolution': key,
+                    'ext': 'mp4',
+                    'acodec': 'AAC',
+                    'vcodec': 'H264',
+                    'url': video_url
+                })
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:137]: Lot of repeated code. Can be refactored into a single block

File: yt_dlp/extractor/smotrim.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        else:
+            raise ExtractorError('Unknown error: ' + meta.get('errors'), expected=False)
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:139]: Use fstring, not +

File: yt_dlp/extractor/smotrim.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+        # sort formats
+        formats = sorted(formats, key=lambda item: int(item['resolution']))
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:142]: self._sort_formats

File: yt_dlp/extractor/smotrim.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        }
+        video_template = 'https://cdn-v.rtr-vesti.ru/_cdn_auth/secure/v/vh/mp4/{quality}/{id_split}.mp4?auth=mh&vid={id}'
+
+        formats = []
+        if meta.get('errors') == '':
+            for key in traverse_obj(meta, ('sources', 'http')):
+                if key in video_quality:
+                    formats.append({
+                        'format': video_quality[key],
+                        'resolution': key,
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:99]: resolution? or height?

File: yt_dlp/extractor/smotrim.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+            raise ExtractorError('Unknown error: ' + meta.get('errors'), expected=False)
+
+        # sort formats
+        formats = sorted(formats, key=lambda item: int(item['resolution']))
+        for i in range(len(formats)):
+            formats[i]['format_id'] = str(i + 1)
+            formats[i]['resolution'] += 'p'
+
+        # import hashlib
+        # print('md5:' + hashlib.md5(meta.get('anons').encode('utf-8')).hexdigest())
```
Comment (pukkandan) [inline yt_dlp/extractor/smotrim.py:148]: ?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/smotrim.py [hunk 17]
```diff
@@ -58,16 +58,17 @@ class SmotrimIE(InfoExtractor):
...
+                r'"https://player.smotrim.ru/iframe/video/id/(?P<video_id>[0-9]+)/',
+                webpage, 'video_id', default=None)
             if video_id is None:
                 raise ExtractorError('This page doesn\'t contain video.', expected=True)
             webpage = self._download_webpage('https://smotrim.ru/video/' + video_id, None, 'Redirect to video')
 
         # Example: https://player.smotrim.ru/iframe/video/id/1539617/start_zoom/true/showZoomBtn/false/sid/smotrim/isPlay/true/mute/true/?acc_video_id=1488382"
         m = re.search(
-            r'"https://player.smotrim.ru/iframe/video/id/(?P<video_id>[0-9]+)/[^"]*?/?acc_video_id=(?P<player_id>[0-9]+)"',
+            r'"https://player.smotrim.ru/iframe/video/id/(?P<video_id>[0-9]+)/[^"]*?/\?acc_video_id=(?P<player_id>[0-9]+)"',
```
Comment (HobbyistDev) [inline yt_dlp/extractor/smotrim.py:71]: Use `\d`

**Comment by Lesmiscore** [comment]
@nikita-moor Are you willing to continue on this PR? If not, I'll be refactoring it

**Comment by nikita-moor** [comment]
> @nikita-moor Are you willing to continue on this PR? If not, I'll be refactoring it (If I don't get response in 12 hours I'll start working on it)

No, I'm not going to elaborate this PR, so you are free to refactor it.

**Pull request closed without merging**